### DISCUSSION TITLE
My question concerns the logic of building a neural network.

### DISCUSSION THREAD

**Discussion by Duwimpuhwe** [mrdbourke/pytorch-deep-learning#484] [Q&A]
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/112572927/fef1ea20-1be6-4be2-a1fd-5a287d6cd8a1)

With created instance of built model, the input shape  is equivalent to the size number of pixels of a flatten image 28*28=784, but during the training and prediction the batch of 32 images is flattened and passed to the model (through the first layer onwards). how does the batch of image processed while the create models has the first layer with neurons matching with the size of the single images?

**Comment by mrdbourke** [comment]
Hi @Duwimpuhwe , 

Good question!

That can be a bit confusing when you start to work with batches.

But the way `torch.nn.Linear` (and many other layers) are designed, they can work with multiple sized inputs on the first dimension (most often the batch dimension).

E.g. `(32, 784)` could be a tensor of batch size 32 of flattened images with size 784.

See the [`torch.nn.Linear` documentation](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) screenshot for more: 

<img width="746" alt="Screenshot 2023-06-15 at 3 57 32 pm" src="https://github.com/mrdbourke/pytorch-deep-learning/assets/16750345/f9c6b115-d839-4318-85f7-8acddabd3072">

I'd also try playing around with some example code/tensors to get a feel of what the inputs/outputs of the layers feel like.

For example: 

```python
import torch
from torch import nn

m = nn.Linear(20, 30)
input = torch.randn(128, 20)
output = m(input)
print(output.size())
```
